Yes, I agree with Larry. The issue I linked to appears to still be relevant. The error message I get is

```
CMake Error at /usr/share/cmake-3.7/Modules/CMakeDetermineSystem.cmake:168 (file):
  file attempted to write a file:
  /home/nils/SpECTRE/test_build/CMakeFiles/CMakeOutput.log into a source
  directory.
Call Stack (most recent call first):
  CMakeLists.txt:7 (project)


CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_Fortran_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
```

which quite frankly I don't even know what it could possibly mean. To make things worse, when I check the output log, it's empty. I think there is a reason why this feature is undocumented.

The other question we should consider is if we want to use a regex to block building everywhere in the source directory, including if someone creates `source/build` and builds in there. I don't do this, but I know other people do, so I can't say if there are advantages or disadvantages to this. I could see git wanting to track the directory, which is bad, though I can understand the convenience of knowing which source and build directory are associated. Overall, I think given that it would allow people to accidentally commit build output into the source tree I'd be opposed to allowing people to do such things.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -7,3 +7,7 @@ cmake_minimum_required(VERSION 3.3.2)
 project(SpECTRE VERSION 0.0.0 LANGUAGES CXX)
 
 set(CMAKE_VERBOSE_MAKEFILE OFF)
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake")
+
+include(ProhibitInSourceBuild)

--- a/cmake/ProhibitInSourceBuild.cmake
+++ b/cmake/ProhibitInSourceBuild.cmake
@@ -1,0 +1,27 @@
+
+# Distributed under the University of Illinois/NCSA Open Source License.
+# See LICENSE.txt for details.
+
+string(REGEX REPLACE "^${CMAKE_SOURCE_DIR}/([^/]*)/?.*$" "\\1"
+  SUBDIR_OF_SOURCE_DIR ${CMAKE_BINARY_DIR})
+set(PROHIBITED_SUBDIRS "cmake" "docs" "src" "tests" "tools")
+
+if ("${CMAKE_BINARY_DIR}" STREQUAL "${CMAKE_SOURCE_DIR}" OR
+    ${SUBDIR_OF_SOURCE_DIR} IN_LIST PROHIBITED_SUBDIRS)
+  message(FATAL_ERROR "\n"
+    "You attempted to build ${PROJECT_NAME} in the directory:\n"
+    "  ${CMAKE_BINARY_DIR}\n"
+    "In-source builds, however, are not allowed. "
+    "Please create a directory and run cmake from there, passing the path "
+    "to the source directory as the last argument; for example:\n"
+    "  cd ${CMAKE_SOURCE_DIR}\n"
+    "  mkdir build\n"
+    "  cd build\n"
+    "  cmake [OPTIONS] ${CMAKE_SOURCE_DIR}\n"
+    "You also need to remove the CMakeCache.txt file and the "
+    "CMakeFiles directory in the source directory, or you will trigger "
+    "this error again, even when doing an out-of-source build. Run:\n"
+    "  rm -r ${CMAKE_BINARY_DIR}/CMakeCache.txt "
+    "${CMAKE_BINARY_DIR}/CMakeFiles\n"
+    )
+endif()

